homeassistant/components/legrandinone.py

```diff
@@ -0,0 +1,459 @@
+    hass.data[DATA_ENTITY_LOOKUP] = {
+        EVENT_KEY_COMMAND: defaultdict(list),
+    }
+
+    # Allow platform to specify function to register new unknown devices
+    hass.data[DATA_DEVICE_REGISTER] = {}
+    hass.data[DATA_DEVICE_REGISTER][EVENT_KEY_COMMAND] = {}
+    hass.data[IOBL_PROTOCOL_HANDLE] = None
+
+    async def async_send_command(call):
+        """Send legrandinone command."""
+        _LOGGER.debug('LegrandInOne command for %s', str(call.data))
+
+        data = {
+            'legrand_id': call.data.get(CONF_DEVICE_ID),
+            'who': call.data.get(CONF_DEVICE_TYPE),
+            'mode': call.data.get(CONF_COMM_MODE),
+            'media': call.data.get(CONF_COMM_MEDIA),
+            'unit': call.data.get(CONF_DEVICE_UNIT),
+            'what': call.data.get(CONF_COMMAND)
+        }
+
+        data['type'] = call.data.get(CONF_PACKET_TYPE)
+        if data['type'] is None:
+            # Default to bus_command packet if not specified.
+            data['type'] = 'bus_command'
+
+        data['values'] = call.data.get(CONF_DIMENSION_VALUES)
+
+        if not await LegrandInOneCommand.send_command(data):
+            _LOGGER.error('Failed LegrandInOne command for %s', str(call.data))
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SEND_COMMAND, async_send_command,
+        schema=SEND_COMMAND_SCHEMA)
+
+    @callback
+    def event_callback(event):
+        """Handle incoming legrandinone events.
+
+        IOBL events arrive as dictionaries of varying content
+        depending on their type. Identify the events and distribute
+        accordingly.
+        """
+        event_type = event.get('type')
+        _LOGGER.debug('event of type %s: %s', event_type, event)
+
+        # Don't propagate non entity events (eg: version string, ack response)
+        if event_type not in hass.data[DATA_ENTITY_LOOKUP]:
+            _LOGGER.debug('unhandled event of type: %s', event_type)
+            return
+
+        # Lookup entities who registered this device id as device id
+        event_id = event.get('legrand_id', None)
+
+#        entity_ids = hass.data[DATA_ENTITY_LOOKUP][event_type][event_id]
+
+#        if entity_ids:
+        if hass.data[DATA_ENTITY_LOOKUP][event_type][event_id]:
+            # Propagate event
+            _LOGGER.debug('passing event to %s', event_id)
+            async_dispatcher_send(hass,
+                                  SIGNAL_HANDLE_EVENT.format(event_id),
+                                  event)
+            return
+
+        device_type = event.get('who')
+
+        # If device is not yet known, register with platform (if loaded)
+        if event_type in hass.data[DATA_DEVICE_REGISTER]:
+            if device_type in hass.data[DATA_DEVICE_REGISTER][event_type]:
+                _LOGGER.debug('device_id not known, adding new device')
+
+                hass.async_create_task(
+                    hass.data[DATA_DEVICE_REGISTER][event_type][
```

Comment:
Here's an example of how to load a platform:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/hlk_sw16.py#L95

Pass any serializable data as discovery_info, which is the fourth argument. If the event is serializable, probably pass the event.

Reply:
So I shall replace the hass.async_create_task(...) by 
hass.async_create_task(async_load_platform(hass,
                                                                       [either 'cover' or 'light' depending on the event],
                                                                        DOMAIN,
                                                                        event,
                                                                        config))
Is that correct ?

Reply:
Yes, if event is serializable.

Reply:
All the config options needed can be passed via discovery_info.

Reply:
Thing is, I'm not sure how to recover from here the platform specific device default config ?

Reply:
In the end, I found a way to overcome this. Might be needed at some point.

Reply:
First we have to move the config options in the platform config schema to the config schema in the component. Look at the rainmachine component eg.